### ISSUE TITLE
Implement a clear and user friendly error message and warning informing users they cannot access V2 endpoints while assigned only the space supporter role.

### ISSUE THREAD

**Issue by monamohebbi** [cloudfoundry/cloud_controller_ng#2434]
#### Issue
Implement a clear and user friendly error message informing users they cannot access V2 endpoints while assigned **only** the space supporter role.

#### Context
We are hoping to avoid untangling the permissions logic in V2 which means the Space Supporter role will not be recognized as a permitted user by any V2 endpoint. This could lead to unexpected or unexplained output if a user with this role were to be using a platform that uses mixed V2 and V3 endpoints. Instead we want to catch any V2 calls early and fail consistently with a helpful error message.

#### Dev Note:
See this [spike branch](https://github.com/cloudfoundry/cloud_controller_ng/commit/a6534f7ed061ba20972889942f3ca559cbc33c75) for inspiration.

#### Expected result
A space supporter with no other roles assigned should fail when curling any endpoint that is prefixed with a v2. The output of this message should inform them why they have failed and tell them to inform their platform admin that their permissions need to be changed. Example:
```
{
  "errors": [
    {
      "code": 10003,
      "detail": "You are not authorized to perform the requested action. See section 'Space Supporter Role in V2' https://docs.cloudfoundry.org/concepts/roles.html",
      "title": "CF-NotAuthorized"
    }
  ]
}
```

A space supporter curling a v2 endpoint with other roles assigned should receive a warning with the following message.
`This endpoint might behaved unpredictably to users assigned the SpaceSupporter role. Check with your admin if you should have this role assigned on a platform using deprecated endpoints. See section 'Space Supporter Role in V2' https://docs.cloudfoundry.org/concepts/roles.html"` 

#### Acceptance
There is a test at the base level that ensures curling endpoints with v2 prefixes as a user with only the space supporter assigned will cause the above error message.

There is a test at the base level that ensures curling endpoints with v2 prefixes as a user with multiple roles including the space supporter assigned will cause the above error message.

#### Documentation
When I browse to any of these endpoints on [valid role types](https://v3-apidocs.cloudfoundry.org/version/3.102.0/index.html#valid-role-types) page of v3 docs, I can see an indication that this role is not recognized by any V2 endpoint, and will fail if a user attempt to curl any v2 endpoint with this role assigned.

When I browse the [offical CF docs](https://docs.cloudfoundry.org/concepts/roles.html) I see information on the space supporter. Additionally, there is a section titled 'Space Supporter Role in V2' with information on why this role does not work in V2 and what steps an admin could take to give their user access or update their platform.

**Comment by weymanf** [comment]
@stephanme @monamohebbi 
Prior to this issue: https://github.com/cloudfoundry/cloud_controller_ng/issues/2385

I know we decided on option #2  to show a helpful error, but there might be performance implications we aren't aware of.

This is how we made it work so far

https://github.com/cloudfoundry/cloud_controller_ng/blob/6592956e53424b8ac60f5fcf13c6af55bea3dad1/middleware/block_unsupported_roles.rb#L25-L32

We have a middleware they will sit in front of all endpoints that are not '/v3', and it will try to make a roles query as long as the user has a valid token. 

The role model currently is a [giant union table](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/app/models/runtime/role.rb). I don't if making this query for every `/v2` call is justifies letting some users silently get 403s.

Any thoughts? 
These other options are still on the table.

Option 1 of failing all space_supporters definitely is bad behavior but it gives us better performance.
Option 3 will be inconsistent in behavior, and confuse users before they realize they have to read our roles docs. But we wouldn't have to write any code.

**Comment by stephanme** [comment]
Thanks for bringing this topic up. Performance is very important for us (SAP) in our very large landscapes and v2 is still under heavy usage (~50% of all CF API requests).

Some table size numbers for a big foundation:

spaces_application_supporters    0 (not yet enabled)
spaces_auditors                75k
spaces_developers             308k 
spaces_managers               151k 


In order to implement option 2 efficiently, it could help to move most of the load to the DB, e.g. by this query:

```
select 'SpaceManager' as type, count(*) from spaces_managers where user_id=?
union all 
select 'SpaceDeveloper', count(*) from spaces_developers where user_id=?
union all 
select 'SpaceAuditor', count(*) from spaces_auditors where user_id=?
union all 
select 'SpaceSupporter', count(*) from spaces_application_supporters where user_id=?;
```

which returns directly the data needed by the middleware. This avoids loading all the space roles of the user (good 90 for myself):

```
      type      | count
----------------+-------
 SpaceSupporter |     0
 SpaceAuditor   |    89
 SpaceManager   |     1
 SpaceDeveloper |     3
(4 rows)
```
Stmt execution time is < 30ms on our (very big) PosteSQL instance.

Another option to address performance concerns would be to introduce a feature flag (capi-release config parameter) to enable/disable this check which is effectively a selection between option 2 and option 3.

**Comment by weymanf** [comment]
Do we not care if they have any org manager or billing roles?

**Comment by stephanme** [comment]
Yes, we want the org roles as well considered:

Role table sizes:

spaces_application_supporters 0 (not yet enabled)
spaces_auditors 75k
spaces_developers 308k
spaces_managers 151k
organizations_auditors 53k     
organizations_billing_managers 50
organizations_managers 260k
organizations_users 343k

The union statement to count role assignments:

```
select 'SpaceManager' as type, count(*) from spaces_managers where user_id=?
union all 
select 'SpaceDeveloper', count(*) from spaces_developers where user_id=?
union all 
select 'SpaceAuditor', count(*) from spaces_auditors where user_id=?
union all 
select 'SpaceSupporter', count(*) from spaces_application_supporters where user_id=?
union all 
select 'OrgManager', count(*) from organizations_managers where user_id=?
union all 
select 'OrgBillingManager', count(*) from organizations_billing_managers where user_id=?
union all 
select 'OrgAuditor', count(*) from organizations_auditors where user_id=?
union all 
select 'OrgUser', count(*) from organizations_users where user_id=?;
```

Sample result:
```
       type        | count
-------------------+-------
 OrgBillingManager |     0
 SpaceSupporter    |     0
 OrgAuditor        |    37
 SpaceAuditor      |    88
 SpaceManager      |     0
 OrgManager        |     0
 OrgUser           |    37
 SpaceDeveloper    |     2
(8 rows)
```

OrgUser role has some special v2/v3 behavior according to https://docs.cloudfoundry.org/concepts/roles.html: it is granted automatically to a user when any other org or space role is assigned via the v2 API (but not when assigned via v3).
But you can't assign SpaceSupporter role via v2 and this way bypass 'option 2'. So it should not contribute to more v2/v3 confusion, at least not for end users.

Execution time of the big union stmt was < 40ms. Analyse on PostgreSQL shows a mixture of index and table scans. There is no index on user_id, only on on org/space_id + user_id. I would say that execution will grow linearly with the size of the roles tables.

+1 for a configuration flag to switch the check off (= select option 3).

If option 2 gets implemented, it is a good candidate for a performance test as part of https://github.com/cloudfoundry-incubator/cf-performance-tests. We could measure e.g. the impact of # of role assignments on a e.g. ``/v2/apps/:guid``.

**Comment by weymanf** [comment]
@stephanme We're still working on making this check work, and we want to bench mark this query translated from the `Sequel` library. We don't have as a big of a data set so could y'all possibly run some performance testing on this query:

```
"SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT 'space_manager' AS \"type\" FROM \"spaces_managers\" WHERE (\"user_id\" = 1) UNION (SELECT 'space_developer' AS \"type\" FROM \"spaces_developers\" WHERE (\"user_id\" = 1))) AS \"t1\" UNION (SELECT 'space_auditor' AS \"type\" FROM \"spaces_auditors\" WHERE (\"user_id\" = 1))) AS \"t1\" UNION (SELECT 'space_supporter' AS \"type\" FROM \"spaces_supporters\" WHERE (\"user_id\" = 1))) AS \"t1\" UNION (SELECT 'organization_manager' AS \"type\" FROM \"organizations_managers\" WHERE (\"user_id\" = 1))) AS \"t1\" UNION (SELECT 'organization_billing_manager' AS \"type\" FROM \"organizations_billing_managers\" WHERE (\"user_id\" = 1))) AS \"t1\" UNION (SELECT 'organization_auditor' AS \"type\" FROM \"organizations_auditors\" WHERE (\"user_id\" = 1))) AS \"t1\""
```

It looks a little crazy but our branch should have the ruby code for it.
It eventually returns a list of roles, and removes the duplicates due to the unions.

**Comment by weymanf** [comment]
We seeded the database with around 139870 roles.

And have two users:

1. Developer - has about 6000 roles
1. Supporter - has 1 to 2 roles
Duration is in seconds

```
Duration: 0.00734575
User: developer
```
```
Duration: 0.00106400
User: supporter
```

Looks like the actual query itself is quick, but it does scale up with the amount of roles a single user has. It doesn't look like the query is a huge a bottleneck.


 Now we just need to find out where to put the check.

**Comment by stephanme** [comment]
Adding @FloThinksPi as I'm currently out of office.

**Comment by FloThinksPi** [comment]
Hey @weymanf are you aware of https://github.com/cloudfoundry-incubator/cf-performance-tests, we made it specifically to test endpoint query times with different db sizes(also large ones) and tracking this info over multiple CC versions to see regressions. Sounds like a nice place to place such a test :)

**Comment by sweinstein22** [comment]
Closed with PR #2467
